typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+* `ErrorCode`
+* Non-pointer CHERI capability
+
+## Type descriptors
+
+When a process wants to send data to the kernel, the process needs a way to tell
+the kernel the sent data's type (and vice versa when the kernel sends data to
+the process). To do this, we need a way to serialize information about a a list
+of types. To start, lets assign numbers to each type (DNE means this type
+doesn't exist yet):
+
+| ID       | Description                  | Kernel Type     | `libtock-c` Type | `libtock-rs` Type         |
+| -------- | ---------------------------- | --------------- | ---------------- | ------------------------- |
+| `0b0001` | Error code                   | `ErrorCode`     | DNE              | `ErrorCode`               |
+| `0b0010` | `u32`                        | `u32`           | `uint32_t`       | `u32`                     |
+| `0b0011` | `i32`                        | `i32`           | `int32_t`        | `i32`                     |
+| `0b0100` | `usize`                      | `usize`         | `size_t`         | `usize`                   |
+| `0b0101` | `isize`                      | `isize`         | `ptrdiff_t`      | `isize`                   |
+| `0b0110` | `u64`                        | `u64`           | `uint64_t`       | `u64`                     |
+| `0b0111` | `i64`                        | `i64`           | `int64_t`        | `i64`                     |
+| `0b1000` | `f32`                        | `f32`           | `float`          | `f32`                     |
+| `0b1001` | `f64`                        | `f64`           | `double`         | `f64`                     |
+| `0b1010` | `bool`                       | `bool`          | `bool`           | `bool`                    |
+| `0b1011` | Upcall pointer               | `CapabilityPtr` | `UpcallFn`       | `UpcallFn`                |
+| `0b1100` | Data pointer                 | `CapabilityPtr` | `T*`             | `*mut T` where `T: Sized` |
+| `0b1101` | Non-pointer CHERI capability | DNE             | DNE              | DNE                       |
+| `0b1110` | *Reserved for future use*    | -               | -                | -                         |
+| `0b1111` | Arbitrary register value     | DNE             | DNE              | `Register`                |
+
+We can describe a list of N types as a 4N bit integer by embedding the Nth type
+ID in the Nth nibble of the integer. So:
+
+* The empty list has ID `0`
+* A list of one type has an ID equal to that type.
+* A list of two types has the first type in bits 0-3 and the second in bits 4-7.
+* A list of three types has the first type in bits 0-3, the second in bits 4-7,
+  and the third in bits 8-11.
+* and so on
+
+For example, the type `(bool, u32, *mut T)` would be described by
+`0b110000101010`, expanded here:
+
+| Bits | Value    | Type           |
+| ---- | -------- | -------------- |
+| 0-3  | `0b1010` | `bool`         |
+| 4-7  | `0b0010` | `u32`          |
+| 8-11 | `0b1100` | Narrow pointer |
```

Comment:
You should state here that the first type goes in the least-significant bits of the resulting value. I.e., the types are stored in little-endian.

Reply:
I explained the same thing three times and still left ambiguity, oops. I added a note that bits 0-3 are the least significant bits.